src/main/java/org/ohdsi/authenticator/service/AbstractInvalidatableTokenProvider.java

```diff
@@ -0,0 +1,35 @@
+package org.ohdsi.authenticator.service;
+
+import io.jsonwebtoken.Claims;
+import org.apache.commons.lang3.StringUtils;
+import org.ohdsi.authenticator.exception.AuthenticationException;
+
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArraySet;
+
+public abstract class AbstractInvalidatableTokenProvider implements TokenProvider {
+
+    protected static final String INVALID_TOKEN_ERROR = "Expired or invalid JWT token";
+    public static final String EMPTY_TOKEN_ERROR = "Empty token";
+    protected Set<String> invalidatedTokens = new CopyOnWriteArraySet<>();
+
+    @Override
+    public void invalidateToken(AccessToken token) {
+        invalidatedTokens.add(token.getValue());
+    }
+
+    @Override
+    public Claims validateTokenAndGetClaims(AccessToken token) {
```

Comment:
Just my opinion: You can get rid of abstract class and simplify calls chain by 
renaming validateAndResolveClaimsInternal to validateAndResolveClaims and calling validateToken logic in the beginning of each implementation.

(I do not like protected in java as it is still part of object public api)

Reply:
you don't like template method pattern :) ok I will do this